src/ILCompiler.Compiler/src/IL/Stubs/StartupCode/StartupCodeMainMethod.cs

```diff
@@ -89,9 +89,20 @@ public override MethodIL EmitIL()
                 codeStream.Emit(ILOpcode.call, emitter.NewToken(startup.GetKnownMethod("GetMainMethodArguments", null)));
             }
             codeStream.Emit(ILOpcode.call, emitter.NewToken(_mainMethod));
+
+            // If the main method returns void, return the latched exit code set by Environment.ExitCode.
             if (_mainMethod.Signature.ReturnType.IsVoid)
             {
-                codeStream.EmitLdc(0);
+                MetadataType environmentAugments = Context.SystemModule.GetType("Internal.Runtime.Augments", "EnvironmentAugments");
```

Comment:
Is your concern about putting `GetLatchedExitCode` on `StartupCodeHelpers`? I can do that.

Or are you asking me to design a solution that will work for `AppDomain::ProcessExit`, be compatible with Project N (that doesn't include `StartupCodeHelpers.cs`), be compatible with `Environment.ExitProcess`, and `Test.CoreLib`? If that's the case, I'll rebase this out of this pull request and open an issue.